cryptography/src/secp256r1.rs

```diff
@@ -0,0 +1,672 @@
+//! secp256r1 implementation of the `Scheme` trait.
+
+use crate::{PrivateKey, PublicKey, Scheme, Signature};
+use commonware_utils::union_unique;
+use p256::{
+    ecdsa::{
+        signature::{Signer, Verifier},
+        SigningKey, VerifyingKey,
+    },
+    elliptic_curve::scalar::IsHigh,
+};
+use rand::{CryptoRng, Rng, SeedableRng};
+use std::borrow::Cow;
+
+const PUBLIC_KEY_LENGTH: usize = 33;
+const SIGNATURE_LENGTH: usize = 64;
+
+#[derive(Clone)]
+pub struct Secp256r1 {
+    signing_key: SigningKey,
+    verifying_key: VerifyingKey,
+}
+
+impl Scheme for Secp256r1 {
+    fn new<R: CryptoRng + Rng>(r: &mut R) -> Self {
+        let signing_key = SigningKey::random(r);
+        let verifying_key = signing_key.verifying_key().to_owned();
+        Self {
+            signing_key,
+            verifying_key,
+        }
+    }
+
+    fn from(private_key: PrivateKey) -> Option<Self> {
+        let signing_key = match SigningKey::from_slice(private_key.as_ref()) {
+            Ok(key) => key,
+            Err(_) => return None,
+        };
+        let verifying_key = signing_key.verifying_key().to_owned();
+        Some(Self {
+            signing_key,
+            verifying_key,
+        })
+    }
+
+    fn from_seed(seed: u64) -> Self {
+        let mut rng = rand::rngs::StdRng::seed_from_u64(seed);
+        Self::new(&mut rng)
+    }
+
+    fn private_key(&self) -> PrivateKey {
+        self.signing_key.to_bytes().to_vec().into()
+    }
+
+    fn public_key(&self) -> PublicKey {
+        self.verifying_key.to_encoded_point(true).to_bytes().into()
+    }
+
+    fn sign(&mut self, namespace: Option<&[u8]>, message: &[u8]) -> Signature {
+        let sig: p256::ecdsa::Signature = match namespace {
+            Some(namespace) => self.signing_key.sign(&union_unique(namespace, message)),
+            None => self.signing_key.sign(message),
+        };
+        let normalized = match sig.normalize_s() {
+            Some(normalized) => normalized,
+            None => sig,
+        };
+        normalized.to_vec().into()
+    }
+
+    fn validate(public_key: &PublicKey) -> bool {
+        VerifyingKey::from_sec1_bytes(public_key).is_ok()
+    }
+
+    fn verify(
+        namespace: Option<&[u8]>,
+        message: &[u8],
+        public_key: &PublicKey,
+        signature: &Signature,
+    ) -> bool {
+        let ecdsa_signature = match p256::ecdsa::Signature::from_slice(signature) {
+            Ok(sig) => sig,
+            Err(_) => return false,
+        };
+        if ecdsa_signature.s().is_high().into() {
```

Comment:
nit: indicate `is_high` is `>=` not `>` (function name implies `>` imo)

Reply:
looking at the Trait comments I feel like they are contradicting.
Trait general comment says
```
/// Is this scalar greater than n / 2?
///
/// # Returns
///
/// - For scalars 0 through n / 2: `Choice::from(0)`
/// - For scalars (n / 2) + 1 through n - 1: `Choice::from(1)`
```

While function says:
```
/// Is this scalar greater than or equal to n / 2?
fn is_high(&self) -> Choice;
```

In k256 (did not see tests in p256) I see test indicating `is_high()` returns `true` when Scalar is `>`
https://github.com/RustCrypto/elliptic-curves/blob/7a71e403e49fbe92d6b2ae8fe3eabbfdef124975/k256/src/arithmetic/scalar.rs#L888

Reply:
`looking at the Trait comments I feel like they are contradicting.`

oof lol. good catch.

Reply:
Their code clearly has `>`: https://github.com/RustCrypto/elliptic-curves/blob/7a71e403e49fbe92d6b2ae8fe3eabbfdef124975/p256/src/arithmetic/scalar.rs#L403

Maybe worth making a PR there...

Reply:
The code is actually "correct" as far as the BIP is concerned: https://github.com/bitcoin/bips/blob/master/bip-0062.mediawiki#low-s-values-in-signatures

> The value S in signatures must be between 0x1 and 0x7FFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF 5D576E73 57A4501D DFE92F46 681B20A0 (inclusive).

I'm guessing that `s=0` is not allowed. So, they should probably just fix the comment on the function.

Reply:
> The code is actually "correct" as far as the BIP is concerned: https://github.com/bitcoin/bips/blob/master/bip-0062.mediawiki#low-s-values-in-signatures
> 
> > The value S in signatures must be between 0x1 and 0x7FFFFFFF FFFFFFFF FFFFFFFF FFFFFFFF 5D576E73 57A4501D DFE92F46 681B20A0 (inclusive).
> 
> I'm guessing that `s=0` is not allowed. So, they should probably just fix the comment on the function.

Looks like it should indeed be `(0,n)`: https://github.com/ethereum/RIPs/blob/master/RIPS/rip-7212.md#required-checks-in-verification

(Let's make sure we have a test vector that fails to parse `s` if it is 0, if you don't already?)

Reply:
> I'm guessing that s=0 is not allowed. So, they should probably just fix the comment on the function.

We should verify that and add a test vector either way (if we don't have one already).

Reply:
> > I'm guessing that s=0 is not allowed. So, they should probably just fix the comment on the function.
> 
> We should verify that and add a test vector either way (if we don't have one already).

In FIPS 186-5 (https://csrc.nist.gov/pubs/fips/186-5/final), it says "A pair of integers (r, s), each in the interval [1, n−1]" in 6.4.1.

Found it also says as much in EIP/RIP-7212 (https://github.com/ethereum/RIPs/blob/master/RIPS/rip-7212.md#required-checks-in-verification).

We should make sure we have vectors for this (if we don't already).

Reply:
We should also make sure that we have a test vector for `r=0` or `s = 0` (I would assume we do but worth asking).

In FIPS 186-5 (https://csrc.nist.gov/pubs/fips/186-5/final), it says "A pair of integers (r, s), each in the interval [1, n−1]" in 6.4.1.

The implementation we are using should enforce this for us in Verify.

Reply:
huum, I have not found test vector with `r=0` or `s=0`. I could potentially take a valid test vector and just set `r = 0` but it could just be a misleading test.

Reply:
I think that's alright. Like I clarified elsewhere, I think testing the edge cases (even if they seem trivial) is worth it.

Reply:
(at least is worth it in cryptography land)